src/cargo/ops/cargo_new.rs

```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
 
+enum TemplateDirectory{
+    Temp(TempDir),
+    Normal(PathBuf)
+}
+
+impl TemplateDirectory {
+    fn path(&self) -> &Path {
+        match *self {
+            TemplateDirectory::Temp(ref tempdir) => tempdir.path(),
+            TemplateDirectory::Normal(ref path) => path.as_path()
+        }
+    }
+}
+
+// The type of template we will use.
+#[derive(Debug, Eq, PartialEq)]
+enum TemplateType<'a>  {
+    GitRepo(&'a str),
+    LocalDir(&'a str),
+    Builtin
+}
+
+fn get_template_type<'a>(repo: Option<&'a str>, 
+                         name: Option<&'a str>
+                         ) -> CargoResult<TemplateType<'a>> {
+    match (repo, name) {
+        (Some(repo_url), _) if repo_url.starts_with("http://")  ||
+            repo_url.starts_with("https://") ||
+            repo_url.starts_with("file://") ||
+            repo_url.starts_with("git@") => Ok(TemplateType::GitRepo(repo_url)),
+        (Some(directory), _) => Ok(TemplateType::LocalDir(directory)),
+        (None, Some(_)) => Err(human("A template was given, but no template repository")),
+        (None, None) => Ok(TemplateType::Builtin)
+    }
+}
+
+fn get_input_template(config: &Config, opts: &MkOptions) -> CargoResult<TemplateDirectory> {
+    let path = opts.path;
+    let name = opts.name;
+    let templates_dir = config.template_path();
+    if fs::metadata(&templates_dir).is_err() {
+        try!(fs::create_dir_all(&templates_dir));
+    }
+
+    let template_type = try!(get_template_type(opts.template_repo, opts.template));
+    let template_dir = match template_type {
+        // given template is a remote git repository & needs to be cloned
+        TemplateType::GitRepo(repo_url) => {
+            let template_dir = try!(TempDir::new(name));
+            println!("Checking out repo to dir: {}", template_dir.path().display());
+
+            try!(config.shell().status("Cloning", repo_url));
+            match Repository::clone(repo_url, &*template_dir.path()) {
+                Ok(_) => {},
+                Err(e) => {
+                    return Err(human(format!( 
+                                "Failed to clone repository: {} - {}", path.display(), e)))
+                }
+            };
+            TemplateDirectory::Temp(template_dir)
+        },
+        // given template is a local directory
+        TemplateType::LocalDir(directory) => {
+            TemplateDirectory::Normal(PathBuf::from(directory))
+        },
+        // no template given, use either "lib" or "bin" templates depending on the
+        // presence of the --bin flag.
+        TemplateType::Builtin => {
+            let temp_templates_dir = try!(TempDir::new(name));
+            let temp_templates_path = temp_templates_dir.path().to_path_buf();
+            if opts.bin {
+                try!(create_bin_template(&temp_templates_path));
+            } else {
+                try!(create_lib_template(&temp_templates_path));
+            }
+            TemplateDirectory::Temp(temp_templates_dir)
```

Comment:
If we are using a builtin template, could we avoid writing out to the filesystem? That shouldn't be strictly required, right?

Perhaps the template could get loaded into memory here in all cases?